### PR TITLE
Write/Read positions of the array elements in Portable Serialization [API-1619]

### LINKED ISSUE #894: Fix Portable Array Serializers [API-1619]
Portable array serializers for the following types are different from the Java client, so their implementations should be synchronized:
* Date
* Time
* Timestamp
* TimestampWithTimezone
* Decimal

### PR DESCRIPTION
**Pull request by utku-caglayan**
Fixes #894 

Should effect the following types for arrays:
- Date
- Time
- Timestamp
- TimestampWithTimezone
- Decimal

### CODE REVIEW

File: internal/serialization/default_portable_reader.go [hunk 1]
```diff
@@ -454,6 +465,23 @@ func (pr *DefaultPortableReader) runAtPosition(pos int32, f func()) {
...
 
+func (pr *DefaultPortableReader) readArrayOfTime(f func(input serialization.DataInput) time.Time) (ts []time.Time) {
+	l := pr.input.ReadInt32()
+	if l == nilArrayLength {
+		return
+	}
+	ts = make([]time.Time, l)
+	offset := pr.input.Position()
+	for i := int32(0); i < l; i++ {
+		//pr.input.SetPosition(offset + i*Int32SizeInBytes)
```
Comment (utku-caglayan) [inline internal/serialization/default_portable_reader.go:476]: @yuce this is also possible, I'm not sure about exposing `ObjectDataInput`. Wdyt?

Comment (yuce) [reply internal/serialization/default_portable_reader.go:476]: Let's keep it as is for now.

Comment (utku-caglayan) [reply internal/serialization/default_portable_reader.go]: Removed the comments describing the alternative 👍

**Review by yuce — CHANGES_REQUESTED** [review]
Looks good, here is an idea...

File: internal/serialization/default_portable_writer.go [hunk 2]
```diff
@@ -245,7 +245,19 @@ func (pw *DefaultPortableWriter) WriteDecimal(fieldName string, d *types.Decimal
 
 func (pw *DefaultPortableWriter) WriteDecimalArray(fieldName string, ds []types.Decimal) {
 	pw.setPosition(fieldName, int32(serialization.TypeDecimalArray))
-	WriteDecimalArray(pw.output.ObjectDataOutput, ds)
+	arrLen := len(ds)
+	if arrLen == 0 {
+		pw.output.WriteInt32(nilArrayLength)
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply internal/serialization/default_portable_writer.go]: Nice catch thanks https://github.com/hazelcast/hazelcast-go-client/pull/895/commits/06ab14eb270448bab52adf8cc01bd04210321f1d

File: internal/serialization/default_portable_reader.go [hunk 3]
```diff
@@ -454,6 +465,23 @@ func (pr *DefaultPortableReader) runAtPosition(pos int32, f func()) {
...
+		//pos := pr.input.ReadInt32()
+		pos := pr.input.(*ObjectDataInput).ReadInt32AtPosition(offset + i*Int32SizeInBytes)
```
[the remark this answers is not in the document]
Comment (utku-caglayan) [reply internal/serialization/default_portable_reader.go]: Sure. https://github.com/hazelcast/hazelcast-go-client/pull/895/commits/06ab14eb270448bab52adf8cc01bd04210321f1d

**Review by yuce — APPROVED** [review]
Looks good!

**Pull request merged**